### ISSUE TITLE
java.lang.AbstractMethodError when matching any<Interface>()

### ISSUE THREAD

**Issue by erikhuizinga** [mockk/mockk#271]
<details>

### Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [x] I am running the latest version
- [x] I checked the documentation and found no answer
- [x] I checked to make sure that this issue has not already been filed

</details>

Although I can hardly imagine this issue hasn't been encountered before, I couldn't find the same issue, so I'm opening a new one.

### Expected Behavior

I can match any interface.

### Current Behavior

I get a `java.lang.AbstractMethodError: com.example.Interface$Subclass1.someMethod()V`.

### Failure Information (for bugs)

I'm trying to verify that a method call happened only once on either one of multiple interface mocks. I don't care which of the two, but I do care that the call happened, so I try to `verify(exactly = 1) { any<Interface>().someMethod() }` or `verify(exactly = 1) { or(mock1, mock2).someMethod() }`

#### Steps to Reproduce

Please provide detailed steps for reproducing the issue.

0. Run MWE below.
0. Fail.

#### Context

Please provide any relevant information about your setup. This is important in case the issue is not reproducible except for under certain conditions.

* MockK version: 1.9.1
* OS: macOS 10.14.3 (18D109)
* Kotlin version: 1.3.21
* JDK version: 
   - JRE: 1.8.0_152-release-1248-b01 x86_64
   - JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
* JUnit version: 4.12
* Type of test: unit test OR ~android instrumented test~

#### Failure Logs

N/A

#### Stack trace

```
// -----------------------[ YOUR STACK STARTS HERE ] -----------------------

java.lang.AbstractMethodError: com.example.Interface$Subclass1.someMethod()V

	at com.example.InterfaceMatchTest$verify any Interface#someMethod is called$1.invoke(InterfaceMatchTest.kt:26)
	at com.example.InterfaceMatchTest$verify any Interface#someMethod is called$1.invoke(InterfaceMatchTest.kt:11)
	at io.mockk.impl.eval.RecordedBlockEvaluator$record$block$1.invoke(RecordedBlockEvaluator.kt:24)
	at io.mockk.impl.eval.RecordedBlockEvaluator$enhanceWithNPERethrow$1.invoke(RecordedBlockEvaluator.kt:74)
	at io.mockk.impl.recording.JvmAutoHinter.autoHint(JvmAutoHinter.kt:23)
	at io.mockk.impl.eval.RecordedBlockEvaluator.record(RecordedBlockEvaluator.kt:36)
	at io.mockk.impl.eval.VerifyBlockEvaluator.verify(VerifyBlockEvaluator.kt:30)
	at io.mockk.MockKDsl.internalVerify(API.kt:118)
	at io.mockk.MockKKt.verify(MockK.kt:139)
	at io.mockk.MockKKt.verify$default(MockK.kt:136)
	at com.example.InterfaceMatchTest.verify any Interface#someMethod is called(InterfaceMatchTest.kt:26)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)


// -----------------------[ YOUR STACK TRACE ENDS HERE ] -----------------------
```

#### Minimal reproducible code (the gist of this issue)

```kotlin
// -----------------------[ GRADLE DEFINITIONS ] -----------------------
dependencies {
    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk8:1.3.21"
    testImplementation 'junit:junit:4.12'
    testImplementation "io.mockk:mockk:1.9.1"
}
// -----------------------[ YOUR CODE STARTS HERE ] -----------------------
package com.example

import io.mockk.clearAllMocks
import io.mockk.mockk
import io.mockk.verify
import org.junit.Test

class InterfaceMatchTest {
    private val mock1 = mockk<Interface>(relaxed = true)
    private val mock2 = mockk<Interface>(relaxed = true)

    @Test
    fun `verify Interface#someMethod is called`() {
        /* Error */ verify(exactly = 0) { any<Interface>().someMethod() }

        // Don't try this at home, kids! (don't use non-determinism in unit tests)
        // The call to random() is here, because in my real test case I don't control on which mock
        // instance the call happens, and I don't care:
        // I just want to verify that someMethod() is called
        listOf(mock1, mock2).random().someMethod()

        // I want to verify this...
        /* Error */ verify(exactly = 1) { any<Interface>().someMethod() }
        // ...or this
        /* Error */ verify(exactly = 1) { or(mock1, mock2).someMethod() }

        clearAllMocks()

        // However, I can do this (trivially) perfectly fine
        mock1.someMethod()
        verify(exactly = 1) { mock1.someMethod() }

        // Which shows that the problem probably isn't the matcher, but how the matcher creates some
        // sort of implementation of Interface, without implementing someMethod()  
    }
}

interface Interface {
    fun someMethod()
}
// -----------------------[ YOUR CODE ENDS HERE ] -----------------------
```

**Comment by languitar** [comment]
I suspect this is the same thing

```
java.lang.AbstractMethodError: Receiver class org.neo4j.graphdb.Node$Subclass3 does not define or inherit an implementation of the resolved method abstract getAllProperties()Ljava/util/Map; of interface org.neo4j.graphdb.PropertyContainer.

	at one.plan.neomongoconnector.mappings.User$Companion.fromNode(User.kt:22)
	at one.plan.neomongoconnector.changehandlers.UserGraphChangeHandlerTest$2$1$1.invoke(UserGraphChangeHandlerTest.kt:95)
	at one.plan.neomongoconnector.changehandlers.UserGraphChangeHandlerTest$2$1$1.invoke(UserGraphChangeHandlerTest.kt:22)
	at io.mockk.impl.eval.RecordedBlockEvaluator$record$block$1.invoke(RecordedBlockEvaluator.kt:24)
	at io.mockk.impl.eval.RecordedBlockEvaluator$enhanceWithNPERethrow$1.invoke(RecordedBlockEvaluator.kt:74)
	at io.mockk.impl.recording.JvmAutoHinter.autoHint(JvmAutoHinter.kt:23)
	at io.mockk.impl.eval.RecordedBlockEvaluator.record(RecordedBlockEvaluator.kt:36)
	at io.mockk.impl.eval.EveryBlockEvaluator.every(EveryBlockEvaluator.kt:30)
	at io.mockk.MockKDsl.internalEvery(API.kt:92)
	at io.mockk.MockKKt.every(MockK.kt:104)
	at one.plan.neomongoconnector.changehandlers.UserGraphChangeHandlerTest$2$1.invokeSuspend(UserGraphChangeHandlerTest.kt:95)
	at one.plan.neomongoconnector.changehandlers.UserGraphChangeHandlerTest$2$1.invoke(UserGraphChangeHandlerTest.kt)
	at io.kotlintest.runner.jvm.TestCaseExecutor$executeTest$supervisorJob$1$invokeSuspend$$inlined$map$lambda$1.invokeSuspend(TestCaseExecutor.kt:121)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:32)
	at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:233)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)


java.lang.AbstractMethodError: Receiver class org.neo4j.graphdb.Node$Subclass3 does not define or inherit an implementation of the resolved method abstract getAllProperties()Ljava/util/Map; of interface org.neo4j.graphdb.PropertyContainer.

	at one.plan.neomongoconnector.mappings.User$Companion.fromNode(User.kt:22)
	at one.plan.neomongoconnector.changehandlers.UserGraphChangeHandlerTest$2$1$1.invoke(UserGraphChangeHandlerTest.kt:95)
	at one.plan.neomongoconnector.changehandlers.UserGraphChangeHandlerTest$2$1$1.invoke(UserGraphChangeHandlerTest.kt:22)
	at io.mockk.impl.eval.RecordedBlockEvaluator$record$block$1.invoke(RecordedBlockEvaluator.kt:24)
	at io.mockk.impl.eval.RecordedBlockEvaluator$enhanceWithNPERethrow$1.invoke(RecordedBlockEvaluator.kt:74)
	at io.mockk.impl.recording.JvmAutoHinter.autoHint(JvmAutoHinter.kt:23)
	at io.mockk.impl.eval.RecordedBlockEvaluator.record(RecordedBlockEvaluator.kt:36)
	at io.mockk.impl.eval.EveryBlockEvaluator.every(EveryBlockEvaluator.kt:30)
	at io.mockk.MockKDsl.internalEvery(API.kt:92)
	at io.mockk.MockKKt.every(MockK.kt:104)
	at one.plan.neomongoconnector.changehandlers.UserGraphChangeHandlerTest$2$1.invokeSuspend(UserGraphChangeHandlerTest.kt:95)
	at one.plan.neomongoconnector.changehandlers.UserGraphChangeHandlerTest$2$1.invoke(UserGraphChangeHandlerTest.kt)
	at io.kotlintest.runner.jvm.TestCaseExecutor$executeTest$supervisorJob$1$invokeSuspend$$inlined$map$lambda$1.invokeSuspend(TestCaseExecutor.kt:121)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:32)
	at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:233)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

happens in:
```kotlin
                val mockNode = mockk<Node>()

                mockkStatic(User::class)
                val mockUser = mockk<User>()
                every { User.fromNode(any()) } returns mockUser

                val returnedUser = handler.parseEntity(mockNode)
```

**Comment by shoshonte** [comment]
Any fix or workaround? I am encountering this problem too

**Comment by f9i1n1n1** [comment]
Similar here. Getting `AbstractMethodError` while trying to match a lambda with receiver parameter by the `any()` matcher (or even any kind of custom matcher). Did not find a workaround for days now. Detailed error message: `java.lang.AbstractMethodError: Receiver class kotlin.jvm.functions.Function1$Subclass0 does not define or inherit an implementation of the resolved method abstract invoke(Ljava/lang/Object;)Ljava/lang/Object; of interface kotlin.jvm.functions.Function1.`

**Comment by erikhuizinga** [comment]
I can't judge the imprtance of this issue, but this still is an issue and multiple users have reported here about it.

**Comment by oleksiyp** [comment]
@erikhuizinga  original issue is doing something that is out of MockK possibilities. There is no such construct that can verify one or another call. So you can submit a feature request if you want. As a final resort, you can create a special construct that will `try` first verify and if not succeed `tries` second. So it is kind of OR combinator for `verifies`.

```
try {
  verify(exactly = 1) { mock1.method() }
} catch(ex: Exception) {
  verify(exactly = 1) { mock2.method() }
}
```
 *Note: don't tell me it's ugly... :smile: I know... wrap it... make `fun` of it... but don't complain :smile:*



@languitar this issue you mentioned is not the same and very specific to neo4j I believe. Please open new ticket with full description.

@f9i1n1n1 there is not enough information to judge about your case. Please submit ticket with full description as well.

Closing the issue not to attract with a quite generic description about "AbstractMethodError".
Thanks
